internal/rpcserver/rpcserver.go

```diff
@@ -5849,48 +5849,69 @@ func (s *Server) authMAC(dst, auth []byte) []byte {
 	return dst
 }
 
+// checkAuthHeader checks the HTTP Basic authentication string by comparing
+// it with the already generated hash.
+//
+// The first bool return value signifies auth success (true if successful) and
+// the second bool return value specifies whether the user can change the state
+// of the server (true) or whether the user is limited (false).
+func (s *Server) checkAuthHeader(auth, remoteAddr string) (bool, bool) {
+	mac := make([]byte, 0, sha256.Size)
+	mac = s.authMAC(mac, []byte(auth))
+
+	cmp := subtle.ConstantTimeCompare(mac, s.authsha[:])
+	limitcmp := subtle.ConstantTimeCompare(mac, s.limitauthsha[:])
+	if cmp|limitcmp == 0 {
+		// Request's auth doesn't match either user
+		log.Warnf("RPC authentication failure from %s", remoteAddr)
+		return false, false
+	}
+	return true, cmp == 1
+}
+
+// checkAuthUserPass checks the correctness of username and password by
+// generating the corresponding HTTP Basic authentication string then
+// compare the string with the already generated hash.
+//
+// The first bool return value signifies auth success (true if successful) and
+// the second bool return value specifies whether the user can change the state
+// of the server (true) or whether the user is limited (false).
+func (s *Server) checkAuthUserPass(user, pass, remoteAddr string) (bool, bool) {
+	login := user + ":" + pass
+	auth := "Basic " + base64.StdEncoding.EncodeToString([]byte(login))
+	return s.checkAuthHeader(auth, remoteAddr)
+}
+
 // checkAuth checks the HTTP Basic authentication supplied by a wallet or RPC
 // client in the HTTP request r.  If the supplied authentication does not match
-// the username and password expected, a non-nil error is returned.
+// the username and password expected, will return (false) as the first value.
 //
 // This check is time-constant.
 //
 // The first bool return value signifies auth success (true if successful) and
 // the second bool return value specifies whether the user can change the state
 // of the server (true) or whether the user is limited (false). The second is
 // always false if the first is.
-func (s *Server) checkAuth(r *http.Request, require bool) (bool, bool, error) {
+func (s *Server) checkAuth(r *http.Request, require bool) (bool, bool) {
 	// If admin-level RPC user and pass options are not set, this always
 	// succeeds.  This will be the case when TLS client certificates are
 	// being used for authentication.
 	if s.authsha == ([32]byte{}) {
-		return true, true, nil
+		return true, true
 	}
 
 	authhdr := r.Header["Authorization"]
 	if len(authhdr) == 0 {
 		if require {
 			log.Warnf("RPC authentication failure from %s",
 				r.RemoteAddr)
-			return false, false, errors.New("auth failure")
+			return false, false
```

Comment:
This change is incorrect.  It means the calling code can no longer tell the difference between whether the authentication is required or not.  The Websockets endpoint must allow Websocket clients to authenticate either via HTTP basic authentication **or** via the [authenticate](https://github.com/decred/dcrd/blob/master/docs/json_rpc_api.mediawiki#authenticate) Websocket-only RPC as opposed to the header.

Reply:
The only effective change in this commit is https://github.com/liukun/dcrd/commit/e37ced291766227a32e037bb7b517ddbb5d16f6f . At first I only committed this one line change, then under the suggestions, I added helper functions and a test. No logic changes except the comparison `cmp|limitcmp != 0` is done.

Reply:
Removed this change in the newest commit. Add a test to check this behaviour.